ryan_savill/03_background_subtraction/readme.md

```diff
@@ -0,0 +1,281 @@
+
+
+```python
+from skimage.morphology import white_tophat
+
+# set the radius of the structuring element
+radius = 25
+
+# generate the structuring element
+str_el = disk(radius)
+
+# apply the filter
+scikit_tophat = white_tophat(tribolium, str_el)
+
+# plotting the result
+fig, (ax1, ax2) = plt.subplots(1, 2, figsize =(20,20))
+ax1.imshow(tribolium, cmap = 'gray')
+ax1.set_title('Original')
+ax2.imshow(scikit_tophat, cmap = 'gray')
+ax2.set_title('Scikit Tophat')
+```
+
+
+    
+![png](output_13_1.png)
+    
+
+
+Again we can make a function to shorten the code if we use it more often:
+
+
+```python
+def subtract_background(image, radius=50, light_bg=False):
+    # import libraries
+    from skimage.morphology import white_tophat, black_tophat, disk 
+    
+    # generate structuring element
+    str_el = disk(radius)
+     
+    # use appropriate filter depending on the background colour
+    if light_bg:
+        return black_tophat(image, str_el)
+    else:
+        return white_tophat(image, str_el)
+```
+
+## Detour: Why Functions Are Awesome
+
+Making functions seems a bit tedious at the beginning and it can be much more fun to explore the code without having to make a function everytime, but you will soon get used to writing functions and chosing parameters to put into the functions. Not only that but you can save these functions in a file to use them in other python programs or jupyter notebooks and you will see how much easier this will make coding in python.
+
+I for instance have chosen to work with data in the form of dataframes (if you use [R](https://www.r-project.org/) or [matlab](https://de.mathworks.com/products/matlab.html) this might be familiar to you) with a python library called [pandas](https://pandas.pydata.org/). It comes pre-installed with anaconda and makes my life a little easier as all columns have names unlike numpy arrays, which could create a lot of confusion. When I want to perform some kind of analysis on my dataframes often functions will output results as numpy arrays, so I have to turn them back into dataframes and sometimes the dataframes need to be converted into arrays as input for some libraries. Functions help me everyday, because the small lines of code that convert datatypes or reshape lists for specific functions really sum up if you have to write them everytime. Also these lines of code make the files ineligable, whereas functions are just neat one-liners which I can easily read! But how can you make a library? I'll show you a simple way now and a more in depth way later:
+
+### How to Build Your Own Simple Library
+- First you have to make a python file and name it something recogniseable (like: img_analysis_functs). 
+- Then you have to add the functions you want to include in your library. Make sure that each function also contains all the import statements of the libraries it uses, otherwise you might be met with errors later on! 
```

Comment:
```suggestion
- Thenm you have to add the functions you want to include in your library. Make sure that each function also contains all the import statements of the libraries it uses, otherwise you might be met with errors later on! 
```